{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "35129594",
   "metadata": {},
   "source": [
    "# Load datasets \n",
    "\n",
    "The gene expression tables of UMI counts (cells x genes) for each experiment are found at [GEO:GSE173947](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE173947). Information about each sample (including the experimental conditions, the number of OSNs, the 10x kit, the corresponding raw data at the GEO, etc.) can be found in [Table S1](../data/tables/GSE173947_Table_S1_datasets.csv) and in [GSE173947_Dataset_raw_file_names.csv](../data/tables/GSE173947_Dataset_raw_file_names.csv). \n",
    "\n",
    "As described in the [README](../README.md), please download the `<expt>_umi_counts.csv.gz` files from the experiments of your interest to the [data/raw](../data/raw) folder.\n",
    "\n",
    "This notebooks loads those supplemental `<expt>_umi_counts.csv.gz` files from the GEO and converts the raw UMI counts into `AnnData` objects (saved in [data/processed](../data/processed)) to use for downstream analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9041168",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pprint import pprint\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from osn.preprocess import get_data_folders, find_raw_count_files\n",
    "from osn.preprocess.make_adata import load_expts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2bf977c",
   "metadata": {},
   "source": [
    "## Load supplemental file with information about each dataset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c3c455f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 152 unique samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GEO_sample</th>\n",
       "      <th>genotype</th>\n",
       "      <th>environment</th>\n",
       "      <th>odor</th>\n",
       "      <th>time</th>\n",
       "      <th>source</th>\n",
       "      <th>orig_ident</th>\n",
       "      <th>bam</th>\n",
       "      <th>counts</th>\n",
       "      <th>metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>homecage-1</td>\n",
       "      <td>WT</td>\n",
       "      <td>home-cage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>baseline-cage</td>\n",
       "      <td>baseline-1</td>\n",
       "      <td>homecage-1_possorted_genome_bam.bam</td>\n",
       "      <td>home_cage_umi_counts.csv.gz</td>\n",
       "      <td>home_cage_metadata.csv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>homecage-2</td>\n",
       "      <td>WT</td>\n",
       "      <td>home-cage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>baseline-cage</td>\n",
       "      <td>baseline-10</td>\n",
       "      <td>homecage-2_possorted_genome_bam.bam</td>\n",
       "      <td>home_cage_umi_counts.csv.gz</td>\n",
       "      <td>home_cage_metadata.csv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>homecage-3</td>\n",
       "      <td>WT</td>\n",
       "      <td>home-cage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>baseline-cage</td>\n",
       "      <td>baseline-2</td>\n",
       "      <td>homecage-3_possorted_genome_bam.bam</td>\n",
       "      <td>home_cage_umi_counts.csv.gz</td>\n",
       "      <td>home_cage_metadata.csv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>homecage-4</td>\n",
       "      <td>WT</td>\n",
       "      <td>home-cage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>baseline-cage</td>\n",
       "      <td>baseline-3</td>\n",
       "      <td>homecage-4_possorted_genome_bam.bam</td>\n",
       "      <td>home_cage_umi_counts.csv.gz</td>\n",
       "      <td>home_cage_metadata.csv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>homecage-5</td>\n",
       "      <td>WT</td>\n",
       "      <td>home-cage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>baseline-cage</td>\n",
       "      <td>baseline-4</td>\n",
       "      <td>homecage-5_possorted_genome_bam.bam</td>\n",
       "      <td>home_cage_umi_counts.csv.gz</td>\n",
       "      <td>home_cage_metadata.csv.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>homecage-6</td>\n",
       "      <td>WT</td>\n",
       "      <td>home-cage</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>baseline-cage</td>\n",
       "      <td>baseline-9</td>\n",
       "      <td>homecage-6_possorted_genome_bam.bam</td>\n",
       "      <td>home_cage_umi_counts.csv.gz</td>\n",
       "      <td>home_cage_metadata.csv.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GEO_sample genotype environment odor time         source   orig_ident  \\\n",
       "0  homecage-1       WT   home-cage  NaN  NaN  baseline-cage   baseline-1   \n",
       "1  homecage-2       WT   home-cage  NaN  NaN  baseline-cage  baseline-10   \n",
       "2  homecage-3       WT   home-cage  NaN  NaN  baseline-cage   baseline-2   \n",
       "3  homecage-4       WT   home-cage  NaN  NaN  baseline-cage   baseline-3   \n",
       "4  homecage-5       WT   home-cage  NaN  NaN  baseline-cage   baseline-4   \n",
       "5  homecage-6       WT   home-cage  NaN  NaN  baseline-cage   baseline-9   \n",
       "\n",
       "                                   bam                       counts  \\\n",
       "0  homecage-1_possorted_genome_bam.bam  home_cage_umi_counts.csv.gz   \n",
       "1  homecage-2_possorted_genome_bam.bam  home_cage_umi_counts.csv.gz   \n",
       "2  homecage-3_possorted_genome_bam.bam  home_cage_umi_counts.csv.gz   \n",
       "3  homecage-4_possorted_genome_bam.bam  home_cage_umi_counts.csv.gz   \n",
       "4  homecage-5_possorted_genome_bam.bam  home_cage_umi_counts.csv.gz   \n",
       "5  homecage-6_possorted_genome_bam.bam  home_cage_umi_counts.csv.gz   \n",
       "\n",
       "                    metadata  \n",
       "0  home_cage_metadata.csv.gz  \n",
       "1  home_cage_metadata.csv.gz  \n",
       "2  home_cage_metadata.csv.gz  \n",
       "3  home_cage_metadata.csv.gz  \n",
       "4  home_cage_metadata.csv.gz  \n",
       "5  home_cage_metadata.csv.gz  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ActSeq_24h_umi_counts.csv.gz': 'ActSeq_24h_metadata.csv.gz',\n",
      " 'ActSeq_conc_analog_umi_counts.csv.gz': 'ActSeq_conc_analog_metadata.csv.gz',\n",
      " 'ActSeq_umi_counts.csv.gz': 'ActSeq_metadata.csv.gz',\n",
      " 'Arrb2KO_umi_counts.csv.gz': 'Arrb2KO_metadata.csv.gz',\n",
      " 'ChronicOccl_umi_counts.csv.gz': 'ChronicOccl_metadata.csv.gz',\n",
      " 'OR_swap_umi_counts.csv.gz': 'OR_swap_metadata.csv.gz',\n",
      " 'RevOccl_umi_counts.csv.gz': 'RevOccl_metadata.csv.gz',\n",
      " 'chronic_ACE_umi_counts.csv.gz': 'chronic_ACE_metadata.csv.gz',\n",
      " 'envA_bidirectional_switch_umi_counts.csv': 'envA_bidirectional_switch_metadata.csv',\n",
      " 'envA_timecourse_umi_counts.csv.gz': 'envA_timecourse_metadata.csv.gz',\n",
      " 'env_switch_umi_counts.csv.gz': 'env_switch_metadata.csv.gz',\n",
      " 'home_cage_immature_umi_counts.csv.gz': 'home_cage_immature_metadata.csv.gz',\n",
      " 'home_cage_umi_counts.csv.gz': 'home_cage_metadata.csv.gz',\n",
      " 'opto_umi_counts.csv.gz': 'opto_metadata.csv.gz'}\n",
      "\n",
      "# of samples for each condition:\n",
      "genotype  environment  odor       time                \n",
      "Arrb2KO   envA         cage       5d                      2\n",
      "          home-cage    cage       0h                      2\n",
      "                                  14d                     2\n",
      "OMP-ChR2  home-cage    cage       0h                      2\n",
      "                       ctrl       12h                     3\n",
      "                                  2h                      2\n",
      "                       stimH      2h                      2\n",
      "                       stimL      2h                      2\n",
      "                       stimM      12h                     3\n",
      "                                  2h                      2\n",
      "WT        envA         ACE2h      14d -> 2h               2\n",
      "                       DPG2h      14d -> 2h               2\n",
      "                       DPG30m     14d -> 30m              2\n",
      "                       cage       12h                     2\n",
      "                                  14d                     2\n",
      "                                  24h                     2\n",
      "                                  2h                      4\n",
      "                                  3d                      2\n",
      "                                  45m                     2\n",
      "                                  4h                      2\n",
      "                                  5d                      2\n",
      "          envB         ACE2h      14d -> 2h               2\n",
      "                       DPG2h      14d -> 2h               2\n",
      "                       DPG30m     14d -> 30m              2\n",
      "          home-cage    2HA0_01p   2h                      2\n",
      "                       2HA0_1p    2h                      2\n",
      "                       2HA2h      2h                      2\n",
      "                       4MA2h      2h                      2\n",
      "                       ACE0_01p   2h                      2\n",
      "                       ACE0_1p    2h                      4\n",
      "                                  5d                      2\n",
      "                       ACE10p     5d                      2\n",
      "                       ACE1p      2h                      2\n",
      "                                  5d                      2\n",
      "                       ACE2h      2h                      4\n",
      "                                  5d (open) -> 2h         2\n",
      "                                  5d (unplugged) -> 2h    2\n",
      "                       Aceto 24h  2h + home-cage 22h      2\n",
      "                       DPG        5d                      2\n",
      "                       DPG 24h    2h + home-cage 22h      2\n",
      "                       DPG2h      2h                      4\n",
      "                                  5d (open) -> 2h         3\n",
      "                                  5d (unplugged) -> 2h    3\n",
      "                       DPG30m     30m                     6\n",
      "                       MS2h       2h                      2\n",
      "                       OCT2h      2h                      2\n",
      "                       cage       0h                      4\n",
      "                                  2h                      4\n",
      "Name: orig_ident, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "data_fold = get_data_folders()\n",
    "df_expt = pd.read_csv(data_fold.tables / 'GSE173947_Dataset_raw_file_names.csv')\n",
    "print(f\"Found {df_expt.GEO_sample.nunique()} unique samples\")\n",
    "display(df_expt.head(6))\n",
    "expts = dict(zip(df_expt.counts, df_expt.metadata))\n",
    "pprint(expts)\n",
    "print(\"\\n# of samples for each condition:\")\n",
    "pprint(df_expt.groupby(['genotype', 'environment', 'odor', 'time']).orig_ident.nunique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4ea3712",
   "metadata": {},
   "source": [
    "# Make adata objects from downloaded csv files\n",
    "To run only on the `home_cage` data do:\n",
    "```python\n",
    "load_expts(files, name=\"home_cage\")\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0f550b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ActSeq_conc_analog GSE173947_ActSeq_conc_analog_umi_counts.csv.gz GSE173947_ActSeq_conc_analog_metadata.csv.gz\n",
      "ChronicOccl GSE173947_ChronicOccl_umi_counts.csv.gz GSE173947_ChronicOccl_metadata.csv.gz\n",
      "home_cage GSE173947_home_cage_umi_counts.csv.gz GSE173947_home_cage_metadata.csv.gz\n",
      "env_switch GSE173947_env_switch_umi_counts.csv.gz GSE173947_env_switch_metadata.csv.gz\n",
      "ActSeq GSE173947_ActSeq_umi_counts.csv.gz GSE173947_ActSeq_metadata.csv.gz\n"
     ]
    }
   ],
   "source": [
    "# look to see which csv files were downloaded\n",
    "files = find_raw_count_files(data_fold)\n",
    "for name, (count, meta) in files.items():\n",
    "    print(name, count.name, meta.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c25ff437",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-12-18 12:37:30,189 - INFO - Already found ActSeq_conc_analog_norm.h5ad file for ActSeq_conc_analog, skipping. To overwrite use `force=True`.\n",
      "2021-12-18 12:37:30,190 - INFO - Already found ChronicOccl_norm.h5ad file for ChronicOccl, skipping. To overwrite use `force=True`.\n",
      "2021-12-18 12:37:30,191 - INFO - Already found home_cage_norm.h5ad file for home_cage, skipping. To overwrite use `force=True`.\n",
      "2021-12-18 12:37:30,192 - INFO - Already found env_switch_norm.h5ad file for env_switch, skipping. To overwrite use `force=True`.\n",
      "2021-12-18 12:37:30,192 - INFO - Already found ActSeq_norm.h5ad file for ActSeq, skipping. To overwrite use `force=True`.\n"
     ]
    }
   ],
   "source": [
    "# if you have enough memory could try `low_memory=False` and various `chunksize`\n",
    "load_expts(files, read_kwargs={\"low_memory\":True})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
